### PR TITLE
[RFC] Clean up buffer.c build_stl_str_hl variable declarations

### PR DESCRIPTION
**Pull request by war1025**
I picked this function more or less at random and decided to move the declarations closer to where the variables are used at. 

After further investigation, this is the function that builds the vim statusline.
The code was rather cryptic, so I have fully commented it. The comments may seem excessive, but that is only because after reading the comment it is obvious that the code is doing what the comment says it is. Without the comments it is quite hard to follow.

I have also written a number of tests to cover the basic items supported by statusline. It is by no means a comprehensive test suite for the function.

### CODE REVIEW

File: src/nvim/buffer.c [hunk 1]
```diff
@@ -3062,25 +3040,26 @@ build_stl_str_hl (
...
-      fillable = FALSE;         /* don't change ' ' to fillchar */
-      if (buf_spname(wp->w_buffer) != NULL)
+    {
+      fillable = false;         /* don't change ' ' to fillchar */
+      if (buf_spname(wp->w_buffer) != NULL) {
         STRLCPY(NameBuff, buf_spname(wp->w_buffer), MAXPATHL);
-      else {
-        t = (opt == STL_FULLPATH) ? wp->w_buffer->b_ffname
-            : wp->w_buffer->b_fname;
-        home_replace(wp->w_buffer, t, NameBuff, MAXPATHL, TRUE);
```
Comment (war1025) [inline src/nvim/buffer.c]: the linter told me to change `TRUE` to `true` on this line, but I'm not sure why it complained because I didn't change this line and it was fine before?

Comment (ZyX-I) [inline src/nvim/buffer.c]: @war1025 Linter silences the errors for *blocks* containing *three* lines, line with the error in the middle. Should you change any of the lines in the block you need to also change adjacent lines.

**Comment by justinmk** [comment]
Starting with https://github.com/neovim/neovim/pull/3297 we would like prefer that cleanup PRs include a couple unit tests that cover the function being changed, so we can feel more confident about regressions and future changes / vim patch merges. 

The unit tests _should_ initialize all globals so it won't be as painful as it used to be. 

https://coveralls.io/builds/3555525/source?filename=src%2Fnvim%2Fbuffer.c indicates that `build_stl_str_hl` is not covered by our tests.

**Comment by war1025** [comment]
@justinmk I have this pull pretty well cleaned up. I am looking into writing tests, but I need some info on how to access some things:

I need access to a `win_T*`. `curwin` would work fine for me.
I need to allocate a `char_u*` output buffer. I was thinking maybe `toCstr()` would create one for me if I gave it a string?

I think the other values I can figure out. There are a couple structs at the end of the arguments list, but maybe I can ignore those for now.

**Comment by justinmk** [comment]
>  need access to a win_T*. curwin would work fine for me.

`curwin` is fine.

> I need to allocate a char_u* output buffer. I was thinking maybe toCstr() would create one for me if I gave it a string?

Yes, that should work.

**Comment by war1025** [comment]
Attempted to write unit tests for this function. Not enough things are initialized for it to work in anything but trivial cases. Will try to figure out how the functional tests stuff works.

**Comment by war1025** [comment]
Functional tests are pretty slick...

**Comment by war1025** [comment]
I have done all the work I plan to do on this pull request. I believe someone still needs to go through it before I can mark it was `[RDY]`. I will check back periodically to see if there has been any activity.

File: src/nvim/buffer.c [hunk 2]
```diff
@@ -2785,8 +2785,8 @@ void free_titles(void)
  * If maxwidth is not zero, the string will be filled at any middle marker
  * or truncated if too long, fillchar is used for all whitespace.
  */
-int 
-build_stl_str_hl (
+int
+build_stl_str_hl(
```
Comment (ZyX-I) [inline src/nvim/buffer.c]: I think that since you already touch this you should use style from the style guide:

```C
int build_stl_str_hl(
```

File: src/nvim/buffer.c [hunk 3]
```diff
@@ -2833,15 +2816,16 @@ build_stl_str_hl (
...
-  int maxwid;
-  int zeropad;
-  char_u base;
-  char_u opt;
+
+  typedef enum {
+     DECIMAL     = 10,
+     OCTAL       = 8,
+     HEXIDECIMAL = 16
+  } number_base;
```
Comment (ZyX-I) [inline src/nvim/buffer.c]: It is very strange to see `typedef` inside the function. Also note that according to the style guide this needs to be

```C
typedef enum {
    kNumBaseDecimal = 10,
    kNumBaseOctal = 8,
    kNumBaseHexadecimal = 16,
} NumberBase;
```
```

Comment (war1025) [inline src/nvim/buffer.c]: Do you have a better solution than the enum?

I am fine updating to match the style guide, but if you have an alternate idea I'd also be happy to change it.

Comment (ZyX-I) [inline src/nvim/buffer.c]: @war1025 Why do you need better solution? Just do not place it *inside the function* and correct naming.

Comment (war1025) [inline src/nvim/buffer.c]: I see, your issue was with the enum being inside the function. That is the piece that I missed. I will update.

File: src/nvim/buffer.c [hunk 4]
```diff
@@ -2770,6 +2770,16 @@ void free_titles(void)
 
 # endif
 
+/**
+ * Enumeration specifying the valid numeric bases that can
+ * be used when printing numbers in the status line.
+ **/
```
Comment (ZyX-I) [inline src/nvim/buffer.c]: Style guide accepts *only* `//`-style comments (specifically `///` for doxygen). `/*…*/` would be fine only inside a macros and only as long as they are one-line.

**Comment by war1025** [comment]
@ZyX-I I updated the comments in the function to match the style guide I believe.

**Comment by war1025** [comment]
ping @justinmk @ZyX-I 

I put a good deal of time into this pull. I'd like to see it merged rather than just sitting in review indefinitely. Is there anything I can do to speed the process along?

**Comment by justinmk** [comment]
It won't be indefinite. I will take a look soon.

**Comment by war1025** [comment]
Thanks

File: src/nvim/buffer.c [hunk 5]
```diff
@@ -2770,58 +2770,59 @@ void free_titles(void)
...
-  long num;
-  int width;
-  int itemcnt;
-  int curitem;
   int groupitem[STL_MAX_ITEM];
-  int groupdepth;
   struct stl_item {
+    // Where the item starts in the status line output buffer
     char_u          *start;
+    // The minimum width of the item (at most 50 characters)
```
Comment (justinmk) [inline src/nvim/buffer.c]: better to leave out potentially stale info such as "at most 50"

File: src/nvim/buffer.c [hunk 6]
```diff
@@ -2855,98 +2849,152 @@ build_stl_str_hl (
...
+    // room in our output buffer, exit the loop.
+    if (*fmt_p == NUL || out_p >= out_end_p)
       break;
 
-    /*
-     * Handle one '%' item.
-     */
-    s++;
-    if (*s == NUL)      /* ignore trailing % */
+    // The rest of this loop wil handle a single `%` item.
```
Comment (justinmk) [inline src/nvim/buffer.c]: typo

File: src/nvim/buffer.c [hunk 7]
```diff
@@ -2855,98 +2849,152 @@ build_stl_str_hl (
...
-      t = item[groupitem[groupdepth]].start;
-      *p = NUL;
-      l = vim_strsize(t);
+      // { Determine how long the group is.
+      //   Note: We set the current output position to null
+      //         so `vim_strsize` will work.
+      char_u *t = item[groupitem[groupdepth]].start;
+      *out_p = NUL;
+      long l = vim_strsize(t);
+      // }
```
Comment (justinmk) [inline src/nvim/buffer.c]: Avoid "comment scopes" in most cases. E.g., this comment scope is already fairly clear because of the blank line below. In other cases, a comment on a `while()` is already "scoped" by the `while(){...}` scope.

File: src/nvim/buffer.c [hunk 8]
```diff
@@ -2954,76 +3002,129 @@ build_stl_str_hl (
...
+    bool zeropad = (*fmt_p == '0');
+    if (zeropad) {
+      fmt_p++;
     }
-    if (*s == '-') {
-      s++;
-      l = -1;
+
+    // Denotes that the item should be left-aligned.
+    // This is tracked by using a negative length.
```
Comment (justinmk) [inline src/nvim/buffer.c]: is the "negative length" comment still true? Seems like it is tracked via `left_align` boolean now.

File: src/nvim/buffer.c [hunk 9]
```diff
@@ -3272,212 +3433,362 @@ build_stl_str_hl (
...
+      // number we just printed
+      out_p += STRLEN(out_p);
+
+    // Otherwise, there was nothing to print so mark the item as empty
+    } else {
       item[curitem].type = Empty;
+    }
 
-    if (opt == STL_VIM_EXPR)
+    // Free the string buffer if we allocated it.
```
Comment (justinmk) [inline src/nvim/buffer.c]: this might be a bit obvious :}

File: src/nvim/buffer.c [hunk 10]
```diff
@@ -3272,212 +3433,362 @@ build_stl_str_hl (
...
+
+    // Item processed, move to the next
     curitem++;
   }
-  *p = NUL;
-  itemcnt = curitem;
 
-  if (usefmt != fmt)
+  // Null terminate the output buffer
+  *out_p = NUL;
```
Comment (justinmk) [inline src/nvim/buffer.c]: comment is probably not needed, this is a common idiom in Vim codebase (and C in general I guess)

File: src/nvim/buffer.c [hunk 11]
```diff
@@ -3272,212 +3433,362 @@ build_stl_str_hl (
...
         }
-        /* Fill up for half a double-wide character. */
-        while (++width < maxwidth)
-          *s++ = fillchar;
-      } else
-        s = out + maxwidth - 1;
-      for (l = 0; l < itemcnt; l++)
-        if (item[l].start > s)
+        // Fill up for half a double-wide character.
+        // XXX : This seems impossible given the exit condition above?
```
Comment (justinmk) [inline src/nvim/buffer.c]: Could have been left-over from Vim source before removing `#ifdef` blocks which maybe confused the original author(s). Would be worth peeking at the Vim source here.

File: src/nvim/buffer.c [hunk 12]
```diff
@@ -3272,212 +3433,362 @@ build_stl_str_hl (
...
+      // { Truncate the string
+      char_u *trunc_end_p = trunc_p + trunc_len;
+      STRMOVE(trunc_p + 1, trunc_end_p);
 
-      /* Fill up for half a double-wide character. */
+      // Put a `<` to mark where we truncated at
+      *trunc_p = '<';
+
+      // Advance the pointer to the end of the string
+      trunc_p = trunc_p + STRLEN(trunc_p);
```
Comment (justinmk) [inline src/nvim/buffer.c]: why is it safe to move this out of the `while()` loop below?

File: src/nvim/buffer.c [hunk 13]
```diff
@@ -2855,98 +2849,152 @@ build_stl_str_hl (
...
+      *out_p++ = *fmt_p++;
+      prevchar_isflag = prevchar_isitem = false;
       continue;
     }
-    if (*s == STL_MIDDLEMARK) {
-      s++;
-      if (groupdepth > 0)
+
+    // STL_MIDDLEMARK denotes the separation place between
+    // left and right aligned items.
```
Comment (justinmk) [inline src/nvim/buffer.c]: "denotes the separation place" => "separates". Then this comment fits on 1 line.

File: src/nvim/buffer.c [hunk 14]
```diff
@@ -2855,98 +2849,152 @@ build_stl_str_hl (
...
         continue;
+      }
       groupdepth--;
 
-      t = item[groupitem[groupdepth]].start;
-      *p = NUL;
-      l = vim_strsize(t);
+      // { Determine how long the group is.
+      //   Note: We set the current output position to null
+      //         so `vim_strsize` will work.
```
Comment (justinmk) [inline src/nvim/buffer.c]: this could be inline to the right of `*out_p = NUL;`

File: src/nvim/buffer.c [hunk 15]
```diff
@@ -3088,42 +3205,62 @@ build_stl_str_hl (
...
+      // { Evaluate the expression
+
+      // Store the current buffer number as a string variable
       vim_snprintf((char *)tmp, sizeof(tmp), "%d", curbuf->b_fnum);
       set_internal_string_var((char_u *)"actual_curbuf", tmp);
 
-      o_curbuf = curbuf;
-      o_curwin = curwin;
+      // Switch the curbuf and curwin to the buffer and window we are
+      // evaluating the statusline for.
```
Comment (justinmk) [inline src/nvim/buffer.c]: this is extremely common idiom in Vim source, so should probably leave out this comment.

**Comment by justinmk** [comment]
@war1025 Thanks for the thoughtful and informed changes. 

Other than my 1 technical question and some nudges in the direction of "brevity without sacrificing clarity", it looks good to me. Let me know if you can address the comments and rebase.

File: src/nvim/buffer.c (hunk 11 above)
Comment (war1025) [inline src/nvim/buffer.c]: I looked at the source. Appears just the same there. I'm going to remove it since as far as I can tell it doesn't do anything.

File: src/nvim/buffer.c (hunk 12 above)
Comment (war1025) [inline src/nvim/buffer.c]: In the `while()` loop, the last step is to set `*trunc_p = NUL`, which means on every round except the first, `STRLEN(trunc_p) == 0`, correct?

File: src/nvim/buffer.c (hunk 5 above)
Comment (war1025) [inline src/nvim/buffer.c]: __Done__

File: src/nvim/buffer.c (hunk 6 above)
Comment (war1025) [inline src/nvim/buffer.c]: __Done__

File: src/nvim/buffer.c (hunk 9 above)
Comment (war1025) [inline src/nvim/buffer.c]: I changed the wording a little. 
I mainly want to leave this comment in because the `if()` statement is checking the one place where we did allocate the string buffer. The comment makes it more obvious why we are doing the deallocation only in that case (or potentially other cases in the future).

File: src/nvim/buffer.c (hunk 10 above)
Comment (war1025) [inline src/nvim/buffer.c]: I removed this comment and added one a couple lines down stating more clearly the point I was trying to get across.

```
  // We have now processed the entire statusline format string.
  // What follows is post-processing to handle alignment and
  // highlighting factors.
```

File: src/nvim/buffer.c (hunk 14 above)
Comment (war1025) [inline src/nvim/buffer.c]: __Done__ I removed the scoping. Inline comments are a pet-peeve of mine. So I chose not to make that change.

File: src/nvim/buffer.c (hunk 15 above)
Comment (war1025) [inline src/nvim/buffer.c]: I am a bit on the fence about this one. One of the big issues I have with open source projects vs the codebase where I work is that everything tends to be super under-documented. 

If I am random person wondering how the code works, I have to know a bunch of quirks about the codebase before I can understand anything. 

IMHO if you are changing a global variable to something that isn't what it actually is, that deserves a comment, no matter how common it is. If we don't want the comment, we should refactor things to not rely on the global variables.

That said, if you really want to, I will remove this comment.

**Comment by war1025** [comment]
I went through and made most of the changes you suggested. Please make another pass when you have time.

File: src/nvim/buffer.c (hunk 15 above)
Comment (war1025) [inline src/nvim/buffer.c]: There is a `WITH_BUFFER` macro. It looks like similar commands exist to create a `WITH_WINDOW` macro. I would not be opposed to creating that and using it here.

**Comment by war1025** [comment]
@justinmk I added two macros `WITH_WINDOW` and `WITH_WINDOW_IN_TAB` to hide the switching of `curbuf`, `curwin`, and `curtab`. I updated the places in the codebase that used the `switch_win()` function. I imagine there are many more places that could be updated later.

Please let me know what you think of this change.

**Comment by justinmk** [comment]
@war1025 I really like this approach. I _was_ going to off-handedly suggest using a function or macro in a later PR... this is better than expected :) In either case it could break if there is an early `return`--but at least with this approach the scope is explicit, and could be guarded with a linter...

File: src/nvim/buffer.c [hunk 16]
```diff
@@ -3107,42 +3222,54 @@ build_stl_str_hl (
...
+
+      // Move our position in the output buffer
+      // to the beginning of the expression
+      out_p = t;
+
+      // { Evaluate the expression
 
+      // Store the current buffer number as a string variable
       vim_snprintf((char *)tmp, sizeof(tmp), "%d", curbuf->b_fnum);
       set_internal_string_var((char_u *)"actual_curbuf", tmp);
```
Comment (justinmk) [inline src/nvim/buffer.c]: Amusing side-note: this is a secret "reserved" variable name. In N/Vim, try

    :let g:actual_curbuf='foo' 

with a statusline like:

    :set statusline=%{bufnr('')}

in N/Vim--it will be immediately unlet.

File: src/nvim/buffer.c [hunk 17]
```diff
@@ -3107,42 +3222,54 @@ build_stl_str_hl (
...
 
-      o_curbuf = curbuf;
-      o_curwin = curwin;
-      curwin = wp;
-      curbuf = wp->w_buffer;
+      // Evaluate the expression using the correct window
+      WITH_WINDOW(wp, {
+        // Note: The result stored in `t` is unused.
+        str = eval_to_string_safe(out_p, &t, use_sandbox);
+      });
```
Comment (justinmk) [inline src/nvim/buffer.c]: Sadly, this is not equivalent. `switch_win()` calls `block_autocmds()` which I am assuming is the cause of some weird behavior with this PR. Test case:

```vim
    autocmd! TermOpen * echom 'termopen' localtime()
    function! Foo()
      doautocmd TermOpen
    endfunction
    set statusline=%{Foo()}
```

The autocmd is not triggered by statusline evaluation ("No matching autocmds" is reported every time statusline redraws).

Perhaps `WITH_WINDOW` could take a parameter which prevents disabling of autocmds?

File: src/nvim/buffer.c (hunk 12 above)
Comment (justinmk) [inline src/nvim/buffer.c]: Correct me if I'm wrong: on each loop, `*trunc_p++ = fillchar;` overwrites the previous `NUL`, then the pointer is incremented and its value set to `NUL`. That means `STRLEN(trunc_p)` would be longer than it started with.

File: src/nvim/buffer.c (hunk 16 above)
Comment (ZyX-I) [inline src/nvim/buffer.c]: It is not “secret”: do `helpg actual_curbuf`, you will see a reference under `:h 'statusline'`. But it is missing a tag.

**Comment by war1025** [comment]
@justinmk I have pulled the `WITH_WINDOW` thing out since that looks like it will be a little more work than I initially expected. I removed the comment that started the discussion.

Let me know if there is anything else you will like changed.

**Comment by justinmk** [comment]
@war1025  did you see 
https://github.com/neovim/neovim/pull/3339#discussion_r44507563

File: src/nvim/buffer.c (hunk 12 above)
Comment (war1025) [inline src/nvim/buffer.c]: The string is getting longer after each loop, however we are also moving `trunc_p` a character ahead each loop.

`STRLEN(trunc_p)` counts the number of characters until a null is found starting from `trunc_p`, which at the end of the loop we just set to `NUL`. So `STRLEN(trunc_p)` has to be `0` on each subsequent loop.

**Comment by war1025** [comment]
@justinmk I had not seen that comment. Responded.

File: src/nvim/buffer.c (hunk 12 above)
Comment (oni-link) [inline src/nvim/buffer.c]: This line also needs to check `width+1<maxwidth` before changing `trunc_p`. Otherwise too long tablines could crash nvim.

Comment (justinmk) [inline src/nvim/buffer.c]: Thanks @oni-link . @war1025 can you fix that?
